### ISSUE TITLE
Make enum indexes non-enumerable

### ISSUE THREAD

**Issue by svr93** [microsoft/TypeScript#53973]
# Suggestion

For now TS numeric enums are compiled with indexes so the result can be confusing at runtime:

```
Object.keys(enum_); // "A", "B", 0, 1
```

The behaviour is unusual so many community members just suggest not use enums at all. It’s the key point because **some really useful language feature became “forbidden” because no one understands how to use it correctly**:

1) [TypeScript Enums are Terrible. Here's Why](https://www.reddit.com/r/typescript/comments/yr4vv5/typescript_enums_are_terrible_heres_why/)

2) [Why TypeScript enums suck](https://blog.logrocket.com/why-typescript-enums-suck/)

3) [TypeScript Enums are Bad // Alternatives to use](https://m.youtube.com/watch?v=3aUHskjF7pc&feature=youtu.be)

My idea (not only my - [see](https://github.com/microsoft/TypeScript/issues/4753#issuecomment-579419574)) is to use `enumerable: false` for indexes.

UPD. Some steps for string & numeric enums unification were already made for TS 5.0; the whole part of roadmap concerned the unification:

[Unification Between Numeric and Literal enums](https://github.com/microsoft/TypeScript/pull/50528)

@ahejlsberg wrote some meaningful things:

> The distinction between these two kinds of enums is subtle and has been the source of a fair amount of confusion over time.

## 🔍 Search Terms

enums, enumerable, iteration

## ✅ Viability Checklist

My suggestion meets these guidelines:

- [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code 
- [ ] This wouldn't change the runtime behavior of existing JavaScript code
- [x] This could be implemented without emitting different JS based on the types of the expressions
- [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## Other related issues:

1) [#4753 (Suggestion: Iterate through Enum values)](https://github.com/microsoft/TypeScript/issues/4753)

2) [#2457 (Allow iterating through enum variants with for of)](https://github.com/microsoft/TypeScript/issues/42457)

3) [#51150](https://github.com/microsoft/TypeScript/issues/51150)

4) [#48193 (see the comment)](https://github.com/microsoft/TypeScript/issues/48193#issuecomment-1265703187)

**Comment by jcalz** [comment]
#4753 was already declined, and this is just asking for the same thing in a different way, except more intrusively because it breaks existing behavior.  

I'd love to go back in time to change enums, but in this timeline I think we're probably stuck with this.  If you want to iterate through the forward mappings only you can suppress numeric-like keys yourself:
```ts
enum E { A, B }
console.log(Object.keys(E).filter(k => "" + (+k) !== k)); // "A", "B"
```
(which works, ignoring weirdnesses like #48956).  No, it's not ergonomic, but at least you don't have to wait for TypeScript enums to change.

**Comment by svr93** [comment]
@jcalz
Both solutions will break existing behaviour but my solution is closer to TS Design Goals because for now different JS emitted for different types of enums: indexed keys only added for numeric enums.

#4753 suggests more complicated solution because additional field should be used - [Symbol.iterator]

> I'd love to go back in time to change enums, but in this timeline I think we're probably stuck with this

But what’s the reason? TS 5.0 already broke current enum behaviour and it’s not the first time. Do you see some limiting design decisions that block my suggestion? If so, can you share your thoughts?

> but at least you don't have to wait for TypeScript enums to change.

I strongly disagree with this point because developers listen the “community voice” and don’t use the feature at all. So it leads to more complicated solutions with custom enum implementations; also `const = … as const` noisy constructions are used. I think better to use single solution for real dev cases instead of keeping current non-ergonomic behaviour.

**Comment by jcalz** [comment]
> Both solutions will break existing behaviour

How would making enum objects iterable break people?  Do people depend on enum objects *not* being iterable?   

> #4753 suggests _more complicated solution ... So **it was the reason to decline it**.

This solution is also more complicated because it `Object.defineProperty()` instead of just setting a property.  Am I missing something?

> What’s the reason? Do you see some limiting design decisions that block my suggestion? If so, can you share your thoughts?

I assume that since enumerable reverse mappings in enum objects have been around for quite a while there might well be real world code that enumerates them.  But I don't have any concrete statistics here so maybe it's an acceptable break? 🤷‍♂️ I expect that the TS team will decline this, but I'm just a community member that expects reverse mappings to be present in the key list, and I don't intend to speak on behalf of others, so I'll stop speculating and let them give reasons if they do decline it.

> TS 5.0 already broke current enum behaviour and it’s not the first time.

Cross-referencing to #50528.

> I strongly disagree with this point because developers listen the “community voice” and don’t use the feature at all.

Unfortunately I think that expression-level TypeScript features that don't exist in JavaScript should probably be avoided where possible, including enums (and parameter properties, and I guess `namespace` and... is that it? I guess decorators are off that list now).  If enums are good for TypeScript they'd be good for JavaScript and then there'd be some actual standard behavior that we could use to make decisions.  If there were a time machine and MS sent retrofactor teams back to right past wrongs the result would probably be to destroy enums entirely instead of just enumerable reverse mappings.  

I'll stop interacting now to let you and others discuss if you want.  Good luck!

**Comment by svr93** [comment]
Some thoughts for other community members:

1) > Do people depend on enum objects not being iterable?

The short answer is “no” but I prefer to not guess it because technically adding the new “system” field is breaking change. For example, someone can write lodash-like helper that returns the last element of collection if iterable and returns last indexed value otherwise. We don’t know what people expect for this enum:

```
enum X {
  X = 2,
  Y = 1,
}
```

2) > This solution is also more complicated because it Object.defineProperty() instead of just setting a property. Am I missing something?

I mean “complicated in runtime” when some helpers should be added to prototype chain in case of [Symbol.iterator]

3) > expression-level TypeScript features that don't exist in JavaScript should probably be avoided where possible, including enums

When some functionality already exists and it’s not private then someone will be use it. If enums are bad why not remove it completely (with deprecation step)? Keeping non-ergonomic behaviour is OK only for JS (“don’t break the web”). Namespaces are also not bad-by-default just because JS already has modules. Namespace functionality OK and maybe preferred for some real-world cases (I know at least 2 cases but I don’t wanna to discuss it here).

**Comment by svr93** [comment]
P.S. A good example that TS enums behaviour was changed is my old issue:

[#15591](https://github.com/microsoft/TypeScript/issues/15591)

For now, all my suggestions were applied in different TS versions 😉

**Comment by RyanCavanaugh** [comment]
I don't think this change alone meaningfully moves the needle on whether you should want to use `enum`s or not; there are plenty of other complaints people have which are not even in-principle addressable.

Breaking a ton of working code (I assure you it exists) to change not-wrong behavior that's been shipping for 10 years just isn't a good trade-off.

**Comment by svr93** [comment]
@RyanCavanaugh 
Thanks for your response!
Ok, all my thoughts was presented here so I will not continue the long discussion; I just wanted to show why people really complain:

1) type safety problem (#15591, also you can check @basarat video - https://youtu.be/3aUHskjF7pc?t=115). It was already fixed in TS 5.0
2) lookup problem (current problem - https://youtu.be/3aUHskjF7pc?t=160)

Issues with “unstable” enum indexes and nominal typing I don’t consider as a problem because it’s how enum works in other languages - the enum idea is to provide stable keys, not values.

So, following the great work that was done for enums unification (release 5.0, special thanks to you, @ahejlsberg and @sandersn for working with the PR), some additional changes can be made & it will cover 80-90% problem cases.

Feel free to close it if
